test/jsontest.c

```diff
@@ -0,0 +1,145 @@
+                \"map2\":{\"1\":1,\"3\":2}\
+             },               \
+   \"null\":null              \
+ }";
+static const char json12[] =
+"[           \
+    { \"$binary\":\"abcdef0123456789\" }, \
+    { \"$binary\" :\"AaBbCcDdEdFf\" }, \
+    { \"$binary\":   \"00\" }, \
+    { \"$binary\":\"\" } \
+ ]";
+
+static const char* tests[] = {
+    json0, json1, json2, json3, json4, json5, json6, json7, json8, json9, json10, json11, json12
+};
+
+static const char bad0[] = "abcd";
+static const char bad1[] = "[ 1,2,3,4,5";
+static const char bad2[] = "{ \"foo\":4 ]";
+static const char bad3[] = "[ 1 2 3 4 ]";
+static const char bad4[] = "{ a:b, c:d }";
+static const char bad5[] = "[[[[ 1, 2 ], 3]}]";
+static const char bad6[] = "ab\ncd";
+static const char bad7[] = "] 1,2 [";
+static const char bad8[] = "[ 1,2,+,= ]";
+
+
+static const char* invalid[] = {
+    bad0, bad1, bad2, bad3, bad4, bad5, bad6, bad7, bad8
+};
+
+static uint8_t cbor1[1024];
+static uint8_t cbor2[1024];
+static char json[4096];
+
+int main(int argc, char** argv)
+{
+    DPS_Status status = DPS_OK;
+    int pretty;
+    int i;
+    // Check with and without formatting
+    for (pretty = 0; pretty <= 1; ++pretty) {
+        for (i = 0; i < (sizeof(tests) / sizeof(tests[0])); ++i) {
+            size_t cbor1Len;
+            size_t cbor2Len;
+
+            // Encode as CBOR
+            status = DPS_JSON2CBOR(tests[i], cbor1, sizeof(cbor1), &cbor1Len);
+            if (status != DPS_OK) {
+                goto Failed;
+            }
+            // Decode to JSON
+            status = DPS_CBOR2JSON(cbor1, cbor1Len, json, sizeof(json), pretty);
+            if (status != DPS_OK) {
+                goto Failed;
+            }
+            printf("%s\n", json);
+            // Encode again as CBOR
+            status = DPS_JSON2CBOR(json, cbor2, sizeof(cbor2), &cbor2Len);
+            if (status != DPS_OK) {
+                goto Failed;
+            }
+            // Fidelity check that two CBOR encodings are the same
+            if (cbor1Len != cbor2Len) {
+                printf("Test %d failed: CBOR lengths %zu != %zu\n", i, cbor1Len, cbor2Len);
+                goto Failed;
+            }
+            if (memcmp(cbor1, cbor2, cbor1Len) != 0) {
+                printf("Test %d failed: CBOR encodings are different:\n%s\n\n%s\n\n", i, tests[i], json);
+                goto Failed;
+            }
+        }
+    }
+    // These should all fail
+    for (i = 0; i < (sizeof(invalid) / sizeof(invalid[0])); ++i) {
+        size_t cbor1Len;
+        // Encode as CBOR
+        status = DPS_JSON2CBOR(invalid[i], cbor1, sizeof(cbor1), &cbor1Len);
+        if (status == DPS_OK) {
+            printf("Test of invalid input %d failed\n\n", i);
+            status = DPS_ERR_FAILURE;
+            goto Failed;
+        }
+    }
+
+    printf("Passed\n");
+    return EXIT_SUCCESS;
+
+Failed:
+
+    printf("Failed at test %d %s\n", i, DPS_ErrTxt(status));
+    return EXIT_FAILURE;
+}
```

Comment:
it'd be good to have a json_fuzzer test too, similar to cbor_fuzzer, and also add json_test to test_scripts/run.py

Reply:
Will add to test_scripts/run 
I agree a json_fuzzer test would be good. Will look into that later.